File: src/main/java/fr/ill/ics/cameo/impl/SubscriberImpl.java
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

package fr.ill.ics.cameo.impl;

import com.google.protobuf.InvalidProtocolBufferException;

import fr.ill.ics.cameo.Application;
import fr.ill.ics.cameo.ConnectionTimeout;
import fr.ill.ics.cameo.UnexpectedException;
import fr.ill.ics.cameo.Zmq;
import fr.ill.ics.cameo.Zmq.Context;
import fr.ill.ics.cameo.proto.Messages;
import fr.ill.ics.cameo.proto.Messages.RequestResponse;

public class SubscriberImpl {

	private static final String SYNC = "SYNC";
	private static final String STREAM = "STREAM";
	private static final String ENDSTREAM = "ENDSTREAM";
	private static final String CANCEL = "CANCEL";
	private static final String STATUS = "STATUS";
	
	private ServerImpl server;
	private Zmq.Context context;
	private String url;
	private int publisherPort;
	private int synchronizerPort;
	private Zmq.Socket subscriber;
	private String cancelEndpoint;
	private Zmq.Socket cancelPublisher;
	private String publisherName;
	private int numberOfSubscribers;
	private InstanceImpl instance;
	private boolean endOfStream = false;
	private SubscriberWaitingImpl waiting = new SubscriberWaitingImpl(this);
	
	SubscriberImpl(ServerImpl server, Zmq.Context context, String url, int publisherPort, int synchronizerPort, String publisherName, int numberOfSubscribers, InstanceImpl instance) {
		this.server = server;
		this.context = context;
		this.url = url;
		this.publisherPort = publisherPort;
		this.synchronizerPort = synchronizerPort;
		this.publisherName = publisherName;
		this.numberOfSubscribers = numberOfSubscribers;
		this.instance = instance;
		
		waiting.add();
	}
	
	void init() throws ConnectionTimeout {
		
		// Create the subscriber
		subscriber = context.createSocket(Zmq.SUB);
		
		subscriber.connect(url + ":" + publisherPort);
		subscriber.subscribe(SYNC);
		subscriber.subscribe(STREAM);
		subscriber.subscribe(ENDSTREAM);
		
		// Create an endpoint that should be unique
		cancelEndpoint = "inproc://cancel." + CancelIdGenerator.newId();
		
		// Create a cancel publisher so that it sends the CANCEL message to the status subscriber (connected to 2 publishers)
		cancelPublisher = context.createSocket(Zmq.PUB);
		cancelPublisher.bind(cancelEndpoint);

		// Subscribe to CANCEL
		subscriber.connect(cancelEndpoint);
		subscriber.subscribe(CANCEL);

		// Subscribe to STATUS
		subscriber.connect(instance.getStatusEndpoint());
		subscriber.subscribe(STATUS);
		
		// Synchronize the subscriber only if the number of subscribers > 0
		if (numberOfSubscribers > 0) {
			
			String endpoint = url + ":" + synchronizerPort;
			
			// Create a socket that will be used for several requests.
			RequestSocket requestSocket = server.createSocket(endpoint);
						
			
			
//			// Polling to wait for connection
//			PollItem[] items = { new PollItem(subscriber, ZMQ.Poller.POLLIN) };
//			
//			boolean ready = false;
//			while (!ready) {
//				// The subscriber sends init messages to the publisher that returns SYNC message
//				Zmq.Msg request = server.createInitRequest();
//				Zmq.Msg reply = null;
//				try {
//					reply = requestSocket.request(request);
//					reply.destroy();
//					request.destroy();
//
//				} catch (ConnectionTimeout e) {
//					// do nothing
//				}
//	
//				// Polling until the first SYNC message is received
//				ZMQ.poll(items, 100);
//				
//				if (items[0].isReadable()) {
//					ready = true;
//				}
//			}
			
			
			// polling to wait for connection
			Zmq.Poller poller = new Zmq.Poller(subscriber);
			
			boolean ready = false;
			while (!ready) {
				
				// The subscriber sends init messages to the publisher that returns SYNC message
				Zmq.Msg request = server.createInitRequest();
				Zmq.Msg reply = null;
				try {
					reply = requestSocket.request(request);
					reply.destroy();
					request.destroy();

				} catch (ConnectionTimeout e) {
					// do nothing
				}

				// return at the first response.
				if (poller.poll(100)) {
					ready = true;
				}
			}
			
			// The subscriber is connected and ready to receive data.
			// Notify the publisher that it can send data.
			Zmq.Msg request = server.createSubscribePublisherRequest();
			Zmq.Msg reply = requestSocket.request(request);
			
			byte[] messageData = reply.getFirstData();
			RequestResponse requestResponse = null;
			
			try {
				requestResponse = RequestResponse.parseFrom(messageData);
				
			} catch (InvalidProtocolBufferException e) {
				throw new UnexpectedException("Cannot parse response");
			}

			
			requestSocket.terminate();
		}
	}
	
	public String getPublisherName() { 
		return publisherName;
	}
	
	public String getInstanceName() {
		return instance.getName();
	}
	
	public int getInstanceId() {
		return instance.getId();
	}
	
	public String getInstanceEndpoint() {
		return instance.getEndpoint();
	}
	
	public boolean hasEnded() {
		return endOfStream;
	}
	
	/**
	 * 
	 * @return the byte[] data. If the return value is null, then the stream is finished. 
	 */
	public byte[] receive() {

		while (true) {
			String response = subscriber.recvStr();
			
			if (response.equals(STREAM)) {
				return subscriber.recv();
				
			} else if (response.equals(ENDSTREAM)) {
				endOfStream = true;
				return null;
				
			} else if (response.equals(CANCEL)) {
				return null;
				
			} else if (response.equals(STATUS)) {
				byte[] statusResponse = subscriber.recv();
				
				try {
					Messages.StatusEvent protoStatus = Messages.StatusEvent.parseFrom(statusResponse);
					
					if (instance.getId() == protoStatus.getId()) {
						
						// Get the state
						int state = protoStatus.getApplicationState();
						
						// Test if the state is terminal
						if (state == Application.State.SUCCESS 
								|| state == Application.State.STOPPED
								|| state == Application.State.KILLED
								|| state == Application.State.ERROR) {
							// Exit because the remote application has terminated.
							return null;
						}
					}
					
				} catch (InvalidProtocolBufferException e) {
					throw new UnexpectedException("Cannot parse response");
				}
			}
		}
	}
	
	/**
	 * 
	 * @return the byte[] data. If the return value is null, then the stream is finished. 
	 */
	public byte[][] receiveTwoParts() {

		while (true) {
			String response = subscriber.recvStr();
			
			if (response.equals(STREAM)) {
				byte[][] result = new byte[2][];
				result[0] = subscriber.recv();
				result[1] = subscriber.recv();
				
				return result;
				
			} else if (response.equals(ENDSTREAM)) {
				endOfStream = true;
				return null;
				
			} else if (response.equals(CANCEL)) {
				return null;
				
			} else if (response.equals(STATUS)) {
				byte[] statusResponse = subscriber.recv();
				
				try {
					Messages.StatusEvent protoStatus = Messages.StatusEvent.parseFrom(statusResponse);
					
					if (instance.getId() == protoStatus.getId()) {
						
						// Get the state
						int state = protoStatus.getApplicationState();
						
						// Test if the state is terminal
						if (state == Application.State.SUCCESS 
								|| state == Application.State.STOPPED
								|| state == Application.State.KILLED
								|| state == Application.State.ERROR) {
							// Exit because the remote application has terminated.
							return null;
						}
					}
					
				} catch (InvalidProtocolBufferException e) {
					throw new UnexpectedException("Cannot parse response");
				}
			}
		}
	}
	
	/**
	 * 
	 * @return the string data. If the return value is null, then the stream is finished. 
	 */
	public String receiveString() {
		
		byte[] data = receive();
		
		if (data == null) {
			return null;
		}
		
		return Buffer.parseString(data);
	}
	
	/**
	 * 
	 * @return the int[] data. If the return value is null, then the stream is finished or the data are corrupted. 
	 */
	public int[] receiveInt32() {

		byte[] data = receive();
		
		if (data == null) {
			return null;
		}
		
		return Buffer.parseInt32(data);
	}
	
	/**
	 * 
	 * @return the long[] data. If the return value is null, then the stream is finished or the data are corrupted. 
	 */
	public long[] receiveInt64() {

		byte[] data = receive();
		
		if (data == null) {
			return null;
		}
		
		return Buffer.parseInt64(data);
	}
	
	/**
	 * 
	 * @return the float[] data. If the return value is null, then the stream is finished or the data are corrupted. 
	 */
	public float[] receiveFloat() {

		byte[] data = receive();
		
		if (data == null) {
			return null;
		}
		
		return Buffer.parseFloat(data);
	}

	/**
	 * 
	 * @return the double[] data. If the return value is null, then the stream is finished or the data are corrupted. 
	 */
	public double[] receiveDouble() {

		byte[] data = receive();
		
		if (data == null) {
			return null;
		}
		
		return Buffer.parseDouble(data);
	}
	
	public void cancel() {
	
		cancelPublisher.sendMore(CANCEL);
		cancelPublisher.send("cancel");
	}
	
	public void terminate() {
		
		waiting.remove();
		
		context.destroySocket(subscriber);
		context.destroySocket(cancelPublisher);
	}

	@Override
	public String toString() {
		return "sub." + publisherName + ":" + instance.getName() + "." + instance.getId() + "@" + instance.getEndpoint();
	}
	
}

File: cameo-api-cpp/src/cameo/impl/SubscriberImpl.cpp
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

#include "SubscriberImpl.h"

#include <sstream>

#include "../Serializer.h"
#include "CancelIdGenerator.h"
#include "ServicesImpl.h"
#include "../Server.h"

using namespace std;

namespace cameo {

const std::string SubscriberImpl::SYNC = "SYNC";
const std::string SubscriberImpl::STREAM = "STREAM";
const std::string SubscriberImpl::ENDSTREAM = "ENDSTREAM";
const std::string SubscriberImpl::CANCEL = "CANCEL";
const std::string SubscriberImpl::STATUS = "STATUS";

SubscriberImpl::SubscriberImpl(const Server * server, const std::string & url, int publisherPort, int synchronizerPort, const std::string& publisherName, int numberOfSubscribers, const std::string& instanceName, int instanceId, const std::string& instanceEndpoint, const std::string& statusEndpoint) :
	m_server(server),
	m_publisherName(publisherName),
	m_url(url),
	m_publisherPort(publisherPort),
	m_synchronizerPort(synchronizerPort),
	m_numberOfSubscribers(numberOfSubscribers),
	m_instanceName(instanceName),
	m_instanceId(instanceId),
	m_instanceEndpoint(instanceEndpoint),
	m_statusEndpoint(statusEndpoint),
	m_endOfStream(false) {
}

SubscriberImpl::~SubscriberImpl() {
}

void SubscriberImpl::init() {

	// create a socket for publishing
	m_subscriber.reset(new zmq::socket_t(m_server->m_impl->m_context, ZMQ_SUB));
	m_subscriber->setsockopt(ZMQ_SUBSCRIBE, SYNC.c_str(), SYNC.length());
	m_subscriber->setsockopt(ZMQ_SUBSCRIBE, STREAM.c_str(), STREAM.length());
	m_subscriber->setsockopt(ZMQ_SUBSCRIBE, ENDSTREAM.c_str(), ENDSTREAM.length());
	m_subscriber->setsockopt(ZMQ_SUBSCRIBE, CANCEL.c_str(), CANCEL.length());
	m_subscriber->setsockopt(ZMQ_SUBSCRIBE, STATUS.c_str(), STATUS.length());

	stringstream pubEndpoint;
	pubEndpoint << m_url << ":" << m_publisherPort;

	m_subscriber->connect(pubEndpoint.str().c_str());

	// We must first bind the cancel publisher before connecting the subscriber.
	stringstream cancelEndpoint;

	// We define a unique name
	cancelEndpoint << "inproc://cancel." << CancelIdGenerator::newId();
	m_cancelEndpoint = cancelEndpoint.str();

	m_cancelPublisher = auto_ptr<zmq::socket_t>(new zmq::socket_t(m_server->m_impl->m_context, ZMQ_PUB));
	m_cancelPublisher->bind(m_cancelEndpoint.c_str());

	m_subscriber->connect(m_cancelEndpoint.c_str());
	m_subscriber->connect(m_statusEndpoint.c_str());

	// synchronize the subscriber only if the number of subscribers > 0
	if (m_numberOfSubscribers > 0) {

		stringstream syncEndpoint;
		syncEndpoint << m_url << ":" << m_synchronizerPort;

		string endpoint = syncEndpoint.str();

		// polling subscriber
		zmq_pollitem_t items[1];
		items[0].socket = static_cast<void *>(*m_subscriber);
		items[0].fd = 0;
		items[0].events = ZMQ_POLLIN;
		items[0].revents = 0;

		bool ready = false;

		while (!ready) {

			string strRequestType = m_server->m_impl->createRequest(PROTO_INIT);
			string strRequestData = m_server->m_impl->createInitRequest();
			m_server->m_impl->isAvailable(strRequestType, strRequestData, endpoint, 100);
	
			// wait for 100ms
			int rc = zmq::poll(items, 1, 100);
			if (rc != 0) {
				ready = true;
			}
		}

		m_server->m_impl->subscribeToPublisher(endpoint);
	}
}

bool SubscriberImpl::hasEnded() const {
	return m_endOfStream;
}

bool SubscriberImpl::receiveBinary(std::string& data) {

	while (true) {
		zmq::message_t * message = new zmq::message_t();
		m_subscriber->recv(message);

		string response(static_cast<char*>(message->data()), message->size());
		delete message;

		if (response == STREAM) {
			message = new zmq::message_t();
			m_subscriber->recv(message);
			data = string(static_cast<char*>(message->data()), message->size());
			delete message;

			return true;

		} else if (response == ENDSTREAM) {
			m_endOfStream = true;
			return false;

		} else if (response == CANCEL) {
			return false;

		} else if (response == STATUS) {
			message = new zmq::message_t();
			m_subscriber->recv(message);

			proto::StatusEvent protoStatus;
			protoStatus.ParseFromArray(message->data(), message->size());
			delete message;

			if (protoStatus.id() == m_instanceId) {
				application::State state = protoStatus.applicationstate();

				// test the terminal state
				if (state == application::SUCCESS
					|| state == application::STOPPED
					|| state == application::KILLED
					|| state == application::FAILURE) {
					// Exit because the remote application has terminated.
					return false;
				}
			}
		}
	}

	return false;
}

bool SubscriberImpl::receive(std::string& data) {

	string bytes;
	bool stream = receiveBinary(bytes);

	if (!stream) {
		return false;
	}

	parse(bytes, data);

	return true;
}

bool SubscriberImpl::receive(std::vector<int32_t>& data) {

	string bytes;
	bool stream = receiveBinary(bytes);

	if (!stream) {
		return false;
	}

	parse(bytes, data);

	return true;
}

bool SubscriberImpl::receive(std::vector<int64_t>& data) {

	string bytes;
	bool stream = receiveBinary(bytes);

	if (!stream) {
		return false;
	}

	parse(bytes, data);

	return true;
}

bool SubscriberImpl::receive(std::vector<float>& data) {

	string bytes;
	bool stream = receiveBinary(bytes);

	if (!stream) {
		return false;
	}

	parse(bytes, data);

	return true;
}

bool SubscriberImpl::receive(std::vector<double>& data) {

	string bytes;
	bool stream = receiveBinary(bytes);

	if (!stream) {
		return false;
	}

	parse(bytes, data);

	return true;
}

bool SubscriberImpl::receiveTwoBinaryParts(std::string& data1, std::string& data2) {

	while (true) {
		zmq::message_t * message = new zmq::message_t();
		m_subscriber->recv(message);

		string response(static_cast<char*>(message->data()), message->size());
		delete message;

		if (response == STREAM) {
			message = new zmq::message_t();
			m_subscriber->recv(message);
			data1 = string(static_cast<char*>(message->data()), message->size());
			delete message;

			message = new zmq::message_t();
			m_subscriber->recv(message);
			data2 = string(static_cast<char*>(message->data()), message->size());
			delete message;

			return true;

		} else if (response == ENDSTREAM) {
			m_endOfStream = true;
			return false;

		} else if (response == CANCEL) {
			return false;

		} else if (response == STATUS) {
			message = new zmq::message_t();
			m_subscriber->recv(message);

			proto::StatusEvent protoStatus;
			protoStatus.ParseFromArray(message->data(), message->size());
			delete message;

			if (protoStatus.id() == m_instanceId) {
				application::State state = protoStatus.applicationstate();

				// test the terminal state
				if (state == application::SUCCESS
					|| state == application::STOPPED
					|| state == application::KILLED
					|| state == application::FAILURE) {
					// Exit because the remote application has terminated.
					return false;
				}
			}
		}
	}

	return false;
}

WaitingImpl * SubscriberImpl::waiting() {

	// Waiting gets the cancel publisher.
	return new SocketWaitingImpl(m_cancelPublisher, CANCEL);
}

}